lib/EnzymeCore/src/rules.jl

```diff
@@ -0,0 +1,117 @@
+module EnzymeRules
+
+import EnzymeCore: Annotation
+export Config, ConfigWidth
+export needs_primal, needs_shadow, width, overwritten
+
+import Base: unwrapva, isvarargtype, unwrap_unionall, rewrap_unionall
+
+"""
+    forward(func::Annotation{typeof(f)}, RT::Type{<:Annotation}, args::Annotation...)
+
+Calculate the forward derivative. The first argument `func` is the callable
+for which the rule applies to. Either wrapped in a [`Const`](@ref)), or
+a [`Duplicated`](@ref) if it is a closure.
+The second argument is the return type annotation, and all other arguments are
+the annotated function arguments.
+"""
+function forward end
+
+struct Config{NeedsPrimal, NeedsShadow, Width, Overwritten} end
+const ConfigWidth{Width} = Config{<:Any,<:Any, Width}
+
+needs_primal(::Config{NeedsPrimal}) where NeedsPrimal = NeedsPrimal
+needs_shadow(::Config{<:Any, NeedsShadow}) where NeedsShadow = NeedsShadow
+width(::Config{<:Any, <:Any, Width}) where Width = Width
+overwritten(::Config{<:Any, <:Any, <:Any, Overwritten}) where Overwritten = Overwritten
+
+"""
+	augmented_primal(::Config, func::Annotation{typeof(f)}, RT::Type{<:Annotation}, args::Annotation...)
+
+Must return a tuple of length 2.
+The first-value is primal value and the second is the tape. If no tape is
+required return `(val, nothing)`.
+"""
+function augmented_primal end
+
+"""
+    reverse(::Config, func::Annotation{typeof(f)}, dret::Annotation, tape, args::Annotation...)
+
+Takes gradient of derivative, activity annotation, and tape
+"""
+function reverse end
+
+function _annotate(@nospecialize(T))
```

Comment:
@aviatesk if you could review this code? and `_annotate_tt`. The goal here to generate a signature `Tuple{F, <:Annotation{<:Array}}` as an example